### ISSUE TITLE
关于网络细节

### ISSUE THREAD

**Issue by zhanghongyong123456** [bcmi/CDTNet-High-Resolution-Image-Harmonization#2]
Our paper
![2022-03-16 11-03-54屏幕截图](https://user-images.githubusercontent.com/48466610/158508649-6ec83a05-28fa-4565-9ec6-74ccd805bb3c.png)
ISSAM paper
![2022-03-16 11-04-22屏幕截图](https://user-images.githubusercontent.com/48466610/158508662-25b02e7f-f942-4c20-b7ea-e54bb3712e1c.png)

想请教一下老师，这个像素转换的实现，仅仅包含ISSAM项目中的Encoder以及Decoder（第二张图像的）编解码还是说也包含了前半部分（HRNet + OCR）只是没有画出来，我看咱们论文（第一张图像）只有编解码的（Encoder和Decoder实现），是否只有编解码这部分，没有HRNet+OCR这部分

**Comment by ustcnewly** [comment]
仅包含ISSAM中的ecoder-decoder, HRNet + OCR没有用。我们用的是没有semantic信息的iSSAM。

**Comment by zhanghongyong123456** [comment]
> 仅包含ISSAM中的ecoder-decoder, HRNet + OCR没有用。我们用的是没有semantic信息的iSSAM。

谢谢您的解答，我想请教一下，这个（图1）是如何实现的，我看论文是 平均池化为L维度特征向量连接为2L特征向量f，（1）其中L是多少，（2）我看到合成L维向量时候mask 和F 是交叉计算得到的 ff 和 fb（图二），难道不是分别平均池化吗，是否进行了混合计算（3）特征f 是否经过后续的一个（1x1）卷积操作变成（3,1,1）特征图，因为我看到3DLUT训练分类网络时候，输出是（3，1，1）特征图（图3）
![2022-03-17 18-39-17屏幕截图](https://user-images.githubusercontent.com/48466610/158793750-dea6d051-fd6a-41de-8f55-464625d42bd5.png)

![2022-03-17 18-41-16屏幕截图](https://user-images.githubusercontent.com/48466610/158793787-2bb58ac2-905e-45c2-b16d-4b93f9513b83.png)

![2022-03-17 18-46-13屏幕截图](https://user-images.githubusercontent.com/48466610/158793820-3ecacfc1-e5b6-4143-9093-963839c8b6f5.png)

**Comment by ustcnewly** [comment]
1) L是256;  2) 对前景区域和背景区域分别做average pooling;  3) 需要过一层FC预测LUT的组合系数。

**Comment by zhanghongyong123456** [comment]
> 1. L是256;  2) 对前景区域和背景区域分别做average pooling;  3) 需要过一层FC预测LUT的组合系数。

谢谢您的及时回复

**Comment by zhanghongyong123456** [comment]
> 1. 对前景区域和背景区域分别做average pooling;

老师您好，我好像还是没有明白第二个问题的平均池化操作，因为我看到Fenc特征图是 512维度 1x1特征图，而mask是1维度256x256特征图，如何平均池化分别得到256维度deff以及fb特征图，需要这样计算（图）最后平均池化吗，我早晨的理解是mask  Fenc特征图分别进行平均池化，这样是错误的，不会得到对应的ff以及fb的，

![2022-03-18 17-58-36屏幕截图](https://user-images.githubusercontent.com/48466610/158983200-3c1f3d7e-0b09-4b30-b5e8-5c7404d72a5e.png)，我觉得论文结构图上有交叉计算的部分，而非mask自己做池化得到256维度，Fenc自己池化得到256维度

但是这样操作好像也不太对，因为Fenc早已经不是与原图对应的坐标点了

**Comment by Mia-Cong** [comment]
> > 1. 对前景区域和背景区域分别做average pooling;
> 
> 老师您好，我好像还是没有明白第二个问题的平均池化操作，因为我看到Fenc特征图是 512维度 1x1特征图，而mask是1维度256x256特征图，如何平均池化分别得到256维度deff以及fb特征图，需要这样计算（图）最后平均池化吗，我早晨的理解是mask Fenc特征图分别进行平均池化，这样是错误的，不会得到对应的ff以及fb的，
> 
> ![2022-03-18 17-58-36屏幕截图](https://user-images.githubusercontent.com/48466610/158983200-3c1f3d7e-0b09-4b30-b5e8-5c7404d72a5e.png)，我觉得论文结构图上有交叉计算的部分，而非mask自己做池化得到256维度，Fenc自己池化得到256维度
> 
> 但是这样操作好像也不太对，因为Fenc早已经不是与原图对应的坐标点了

F_enc特征图维度是256不是512，F_enc和下采样的mask相乘分别得到前景和背景，在分别对前景和背景做平均池化就是两个256维，拼接后512维，经过fc得到4个组合系数。

**Comment by zhanghongyong123456** [comment]
> 

恩恩，大佬谢谢及时解疑，我再修改一下

**Comment by zhanghongyong123456** [comment]
> 接后512维，经过fc得到4个组合系数。

不好意思又要打扰你一下：我算了两次得到的Fenc特征图就是 512维度 1x1大小的，这是我的计算过程（按照kernel_size = 4 , stride=2,padding=1），depth默认7：
input = （4，256，256）
block0=（64，128，128）
block1=（64，64，64）
block2=（128，32，32）
block3=（128，16，16）
block4=（256，8，8）
block5=（256，4，4）
block6=（512，1，1）
是我那里算错了吗，这是参照计算代码位置
![2022-03-21 18-30-03屏幕截图](https://user-images.githubusercontent.com/48466610/159243944-abcb777e-ba83-4711-8e03-6bad0a488930.png)

**Comment by Mia-Cong** [comment]
> > 接后512维，经过fc得到4个组合系数。
> 
> 不好意思又要打扰你一下：我算了两次得到的Fenc特征图就是 512维度 1x1大小的，这是我的计算过程（按照kernel_size = 4 , stride=2,padding=1），depth默认7： input = （4，256，256） block0=（64，128，128） block1=（64，64，64） block2=（128，32，32） block3=（128，16，16） block4=（256，8，8） block5=（256，4，4） block6=（512，1，1） 是我那里算错了吗，这是参照计算代码位置 ![2022-03-21 18-30-03屏幕截图](https://user-images.githubusercontent.com/48466610/159243944-abcb777e-ba83-4711-8e03-6bad0a488930.png)

low-resolution generator的depth是4，bottleneck的维度是256

**Comment by zhanghongyong123456** [comment]
> > > 接后512维，经过fc得到4个组合系数。
> > 
> > 
> > 不好意思又要打扰你一下：我算了两次得到的Fenc特征图就是 512维度 1x1大小的，这是我的计算过程（按照kernel_size = 4 , stride=2,padding=1），depth默认7： input = （4，256，256） block0=（64，128，128） block1=（64，64，64） block2=（128，32，32） block3=（128，16，16） block4=（256，8，8） block5=（256，4，4） block6=（512，1，1） 是我那里算错了吗，这是参照计算代码位置 ![2022-03-21 18-30-03屏幕截图](https://user-images.githubusercontent.com/48466610/159243944-abcb777e-ba83-4711-8e03-6bad0a488930.png)
> 
> low-resolution generator的depth是4，bottleneck的维度是256

您好当我设置depth=4，打印出来的bottleneck为[1,128,15,15],而非[1，256，15，15],于是我将ConvEncoder类的里面通道不变模块部分 # if block_i % 2:  设置为 False（图一），这样得到[1,256,15,15]维度的Fenc 特征图，为了和mask[1,1,256,256]进行矩阵计算，我将Fenc进行(256，256)双线性上采样，得到[1,256,256,256],最后计算矩阵得到两个ff、fb为：[1,256,256,256]，最后采用torch.cat合并为[1,512,256,256],在进行一次平均池化得到[1,512,1,1]向量,最后经过fc层得到[1,4]组合系数（图二），请问这样合理吗，期待您的修正。
图一：
![2022-03-22 10-52-20屏幕截图](https://user-images.githubusercontent.com/48466610/159398382-1cc8d031-09cf-4532-ad9a-f77107020254.png)
self.shape = nn.Sequential(nn.Upsample(size=(256,256),mode='bilinear'))   对Fenc进行上采样操作[1，256，15，15]---->[1,256,256,256]
矩阵计算： Fenc  [1,256,256,256]   mask:[1,1,256,256]
分别得到ff fb为 [1, 256, 256, 256]  、[1, 256, 256, 256]
图二：
![2022-03-22 11-03-19屏幕截图](https://user-images.githubusercontent.com/48466610/159399536-b7288e9d-fbfe-4295-8c9b-1211d7b523f5.png)
![2022-03-22 11-07-11屏幕截图](https://user-images.githubusercontent.com/48466610/159399934-169cf8eb-6c61-4fe7-a1e9-1b1ccf24e1ae.png)
